runners/spark/src/main/java/org/apache/beam/runners/spark/SparkRunner.java

```diff
@@ -216,6 +219,34 @@ public CompositeBehavior enterCompositeTransform(TransformTreeNode node) {
       return CompositeBehavior.ENTER_TRANSFORM;
     }
 
+    private boolean shouldDiffer(TransformTreeNode node) {
```

Comment:
Yes, the condition does seem out-of-place, and this was indeed a compromise to get things working, but this is indeed only one exception.
If this turns to be somewhat of a feature the SDK provides (conditional optimization for transformations) we might want to tackle this at `runner-core` level.

Yes, `defer` indeed. 
I'll fix this and merge, Thanks!